nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp

```diff
@@ -0,0 +1,343 @@
+  declareParameter("speed_limit_topic");
+  try {
+    speed_limit_topic_ = node->get_parameter(name_ + "." + "speed_limit_topic").as_string();
+  } catch (rclcpp::exceptions::ParameterNotDeclaredException & ex) {
+    RCLCPP_ERROR(node->get_logger(), "speed_limit_topic parameter is not set");
+    throw ex;
+  }
+
+  filter_info_topic_ = filter_info_topic;
+  // Setting new costmap filter info subscriber
+  RCLCPP_INFO(
+    node->get_logger(),
+    "Subscribing to \"%s\" topic for filter info...",
+    filter_info_topic_.c_str());
+  filter_info_sub_ = node->create_subscription<nav2_msgs::msg::CostmapFilterInfo>(
+    filter_info_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&SpeedFilter::filterInfoCallback, this, std::placeholders::_1));
+
+  // Get global frame required for speed limit publisher
+  global_frame_ = layered_costmap_->getGlobalFrameID();
+
+  // Create new speed limit publisher
+  speed_limit_pub_ = node->create_publisher<nav2_msgs::msg::SpeedLimit>(
+    speed_limit_topic_,
+    rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable());
+  speed_limit_pub_->on_activate();
+
+  // Reset speed conversion states
+  base_ = BASE_DEFAULT;
+  multiplier_ = MULTIPLIER_DEFAULT;
+  percentage_ = false;
+  speed_limit_ = NO_SPEED_LIMIT;
+  speed_limit_prev_ = FAKE_SPEED_LIMIT;  // Fake speed limit to force speed limit topic update
+}
+
+void SpeedFilter::filterInfoCallback(
+  const nav2_msgs::msg::CostmapFilterInfo::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!mask_sub_) {
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Received filter info from %s topic.", filter_info_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      node->get_logger(),
+      "New costmap filter info arrived from %s topic. Updating old filter info.",
+      filter_info_topic_.c_str());
+    // Resetting previous subscriber each time when new costmap filter information arrives
+    mask_sub_.reset();
+  }
+
+  // Set base_/multiplier_ or use speed limit in % of maximum speed
+  if (msg->type == SPEED_FILTER_ABSOLUTE) {
+    // Using speed limit in m/s
+    percentage_ = false;
+    base_ = msg->base;
+    multiplier_ = msg->multiplier;
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Speed filter is using absolute speed_limit = %f + filter_mask_data * %f",
+      base_, multiplier_);
+  } else if (msg->type == SPEED_FILTER_PERCENT) {
+    // Using speed limit in % of maximum speed
+    percentage_ = true;
+    // Checking that base and multiplier are set to their default values
+    if (msg->base != BASE_DEFAULT or msg->multiplier != MULTIPLIER_DEFAULT) {
+      RCLCPP_ERROR(
+        node->get_logger(),
+        "For proper use of speed filter with percentage restriction of maximum speed base and "
```

Comment:
Why can't they be non-default? Maybe they want it that way for some strange reason. I don't think there's any technical reason why we can't support it.

Reply:
The reason is the same as made for KeepoutFilter (https://github.com/ros-planning/navigation2/blob/main/nav2_costmap_2d/plugins/costmap_filters/keepout_filter.cpp#L95): Even though `base` and `multiplier` coefficients are not used by SpeedFilter in percentage case, they should be set to `0.0` and `1.0` accordingly in order to explicitly show that we have one-to-one conversion from OccupancyGrid values -> to a filter value space.

Reply:
In your PR those are set to 

```
static constexpr double BASE_DEFAULT = 0.0;
static constexpr double MULTIPLIER_DEFAULT = 1.0;
```

This `if` statement says that if either these 2 parameters are not set to their default, it will fail to work. What I'm suggesting is they can be anything: `3, 17.3` and should still be processed. Maybe they want to remap 0-100 using a different linear map.

Reply:
The point is flexibility + also these are the 2 major implementations we're going to have for now. So this is going to become a reference design for all future user implemented layers. Having as many examples with using these base / multiplier values is critical to people understanding and being able to fully leverage this work, since documentation alone is never enough.

Reply:
This seems reasonable - to have some working range in percent as well: not always `[0% .. 100%]`, but say `[33.5%..45.6%]`. Good.